### ISSUE TITLE
Speed up rspec specs

### ISSUE THREAD

**Issue by joshsmith** [code-corps/deprecated-code-corps-rails-api#109]
We have a number of really slow specs that could probably be sped up. This is a low-priority refactor. Making a note before we forget.

As of today, here are the slowest:

```
Top 10 slowest examples (12.63 seconds, 23.2% of total time):
  PostSerializer individual resource representation attributes has a 'status'
    1.39 seconds ./spec/serializers/post_serializer_spec.rb:57
  PostSerializer individual resource representation attributes has 'markdown'
    1.29 seconds ./spec/serializers/post_serializer_spec.rb:53
  PostSerializer individual resource representation attributes has a 'title'
    1.29 seconds ./spec/serializers/post_serializer_spec.rb:45
  PostSerializer individual resource representation attributes has a 'body'
    1.29 seconds ./spec/serializers/post_serializer_spec.rb:49
  PostSerializer individual resource representation root has an id
    1.28 seconds ./spec/serializers/post_serializer_spec.rb:30
  PostSerializer individual resource representation root has a type set to 'posts'
    1.26 seconds ./spec/serializers/post_serializer_spec.rb:34
  PostSerializer individual resource representation root has an attributes object
    1.23 seconds ./spec/serializers/post_serializer_spec.rb:26
  PostSerializer individual resource representation attributes has a 'post_type'
    1.22 seconds ./spec/serializers/post_serializer_spec.rb:61
  PostSerializer individual resource representation attributes has a 'likes_count'
    1.18 seconds ./spec/serializers/post_serializer_spec.rb:65
  PostSerializer individual resource representation relationships should include 'project'
    1.18 seconds ./spec/serializers/post_serializer_spec.rb:88

Top 10 slowest example groups:
  PostSerializer
    1.2 seconds average (20.48 seconds / 17 examples) ./spec/serializers/post_serializer_spec.rb:3
  CodeCorps::Scenario::GenerateUserMentionsForComment
    0.30353 seconds average (0.30353 seconds / 1 example) ./spec/lib/code_corps/scenario/generate_user_mentions_for_comment_spec.rb:6
  AddFacebookProfilePictureWorker
    0.23632 seconds average (0.23632 seconds / 1 example) ./spec/workers/add_facebook_profile_picture_worker_spec.rb:3
  PostSerializerWithoutIncludes
    0.23527 seconds average (2.12 seconds / 9 examples) ./spec/serializers/post_serializer_without_includes_spec.rb:3
  ContributorPolicy
    0.19664 seconds average (8.85 seconds / 45 examples) ./spec/policies/contributor_policy_spec.rb:3
  Contributors API
    0.15759 seconds average (2.36 seconds / 15 examples) ./spec/requests/api/contributors_spec.rb:3
  Posts API
    0.1456 seconds average (3.2 seconds / 22 examples) ./spec/requests/api/posts_spec.rb:3
  ProjectSerializer
    0.1297 seconds average (1.43 seconds / 11 examples) ./spec/serializers/project_serializer_spec.rb:3
  CommentUserMentionSerializer
    0.12231 seconds average (1.22 seconds / 10 examples) ./spec/serializers/comment_user_mention_serializer_spec.rb:3
  UpdateProfilePictureWorker
    0.09992 seconds average (0.19984 seconds / 2 examples) ./spec/workers/update_profile_picture_worker_spec.rb:3
```

This list can be regenerated with `rspec spec --profile`.

**Comment by begedin** [comment]
I looked into this a bit. We could speed things up, even signficantly, but at the cost of test readability.

Right now, the API specs are the slowest and this is meanly do tu fragmenting large test we used to do into parts such as
- An API request
  - does something with a record
  - responds with some status code
  - responds with json in a certain format
  - creates some side-effect
  - etc.

In previous projects, we'd just do one big test which asserts all those things. 

The thing is, while the smaller tests are far more readable (and easier to write/maintain individually), the setup (before blocks) runs for each of them once, while in the case of the larger test we used to write, it only ran once.

What we could do 
- merge the smaller tests together again. perhaps rely on a couple of custom matchers to make them more readable even in their big format. it will be waster, but also less readable, and might require some work with matchers, if they make sense at all.
- use `before :all` blocks to speed up some of the stuff, records created this way don't get cleaned up after the test runs, so we have to do manual cleanup in many cases.
- shut up and take it (might still be an option right now, but probably not in the future).

In my opinion, the first option offers the best draw-off, but I'm really leaning towards writing custom matchers in that case. Te `before :all` option needs way too much case-specific micro-management.

**Comment by begedin** [comment]
As a further way to slightly improve readability (and the ability to debug) large tests, we can make use of the `aggregate_failures` option `RSpec` provides.

``` Ruby
it "returns a proper response", aggregate_failures: true do
  expect(last_response.status).to eq 200
  expect(json).to serialize_object(@user).with(AuthenticatedUserSerializer)
end
```

The example test written with this option will output the result of each `expect` that fails, not just the first one.

**Comment by joshsmith** [comment]
Some of this will probably be addressed by @christopherstyles' work on #288.

**Comment by joshsmith** [comment]
Closing as irrelevant because moving to Phoenix.
